### ISSUE TITLE
[explicit-module-boundary-types] Warns on constructor overload

### ISSUE THREAD

**Issue by TheBrokenRail** [typescript-eslint/typescript-eslint#2150]
**Repro**
```typescript
export class Test {
    constructor();
    constrcutor(value?: string) {
        console.log(value);
    }
}
```

**Expected Result**
No warnings

**Actual Result**
```
114:16  warning  Missing return type on function  @typescript-eslint/explicit-module-boundary-types
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.1.0` |
| `@typescript-eslint/parser`        | `3.1.0` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `14.2.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
> Warning on lines 2

Could you please elaborate?

Also this code shouldn't ever match the rule, as it should only inspect exported things - could you please provide a full repro?

**Comment by SkeLLLa** [comment]
I've got this rule warning even without overloading constructor.

```ts
declare namespace MyClass {
    interface Options {
      // ...
    }
}

declare class MyClass {
    constructor(options?: MyClass.Options); // here Missing return type on function  @typescript-eslint/explicit-module-boundary-types
}

export = MyClass;
```

Eslint config:
```js
const tsPlugin = require('@typescript-eslint/eslint-plugin');
module.exports = {
    overrides: [
        {
            files: ['**/*.ts'],
            plugins: ['@typescript-eslint'],
            rules: {
                ...tsPlugin.configs['eslint-recommended'].overrides.rules,
                ...tsPlugin.configs.recommended.rules,
                'no-redeclare': 'off',
                '@typescript-eslint/ban-types': 'off',
                '@typescript-eslint/require-await': 'off',
            },
        },
    ],
};
```

**Comment by TheBrokenRail** [comment]
Updates reproduction!

**Comment by chunghha** [comment]
To me, this warning appears with not only with constructor but also from a function which, both cases, I don't see with v3.0.2. To reproduce, please check out the develop branch of https://github.com/chunghha/docker-ts-nest/tree/develop You will see the warning for former pic below by "yarn lint" with v3.1.0 but won't with v3.0.2 regardless of the eslint-disable-next-line comment there.

![image](https://user-images.githubusercontent.com/24458075/83533520-4ad91d00-a4b5-11ea-96cf-30ae12e27c3d.png)

![image](https://user-images.githubusercontent.com/24458075/83533529-4e6ca400-a4b5-11ea-88c8-c0dc87dbf64a.png)

**Comment by bradzacher** [comment]
@chunghha - this is expected and intentional.
You have parameters on those functions that are _not_ typed.
In the latest release I reworked the error locations so that they highlight the arguments that are untyped.

So your suppression comments are no longer suppressing the errors.

```ts
	public rate<T>(
		@Param('from', new UpperCasePipe()) from: string,
		@Param('to', new UpperCasePipe()) to: string,
		@Res() response
//             ^^^^^^^^ ERROR - THIS PARAMETER IS UNTYPED
	): void {

    constructor(
      @Inject(CONTEXT) private readonly ctx,
//                     ^^^^^^^^^^^^^^^^^^^^ ERROR - THIS PARAMETER IS UNTYPED
      @Inject('winston') private readonly logger: Logger,
      private readonly metadataService: MetadataService
    ) {}
```

**Comment by chunghha** [comment]
@bradzacher - the comment is intentional to suppress the warning because I am aware of some of parameters are hard to be typed. And, prettier does formatting of the lines including line breaks. Therefore, I prefer the behavior of v3.0.2 than v3.1.0 if you refer to this location. But, the fix in v3.1.0 is permanent way forward, I'll then adjust the comment location in my code .
![image](https://user-images.githubusercontent.com/24458075/83552808-f8592a00-a4cf-11ea-89f0-7cfde2696f90.png)

**Comment by bradzacher** [comment]
This is intentional and will be the way forward.
Instead of highlighting the entire function when there is an untyped argument, it's much clearer to just highlighting just the untyped argument.

This makes it easier to find and fix the problem.

You should not leave untyped arguments in the codebase. Untyped arguments are simply typed as `any`, which removes all type safety from the function body, as well as from the function callsites.

**Comment by TheBrokenRail** [comment]
In my issue all the arguments are typed, and it only highlights the overloads.

**Comment by TheBrokenRail** [comment]
It specifically mentions missing the return type.
